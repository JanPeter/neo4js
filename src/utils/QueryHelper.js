// @flow

import { forIn } from "lodash";
import { CharGenerator } from "./CharGenerator";

const singlePredicates = {
  $sw: "STARTS WITH",
  $ew: "ENDS WITH",
  $contains: "CONTAINS",
  $reg: "=~",

  $eq: "=",
  $in: "IN",

  $gt: ">",
  $gte: ">=",
  $lt: "<",
  $lte: "<=",
};

function _prepareWhere(
  props: any,
  variable: string
): { where: string[], flatProps: any } {
  const where = [];
  let flatProps = {};

  if (!props) {
    return { where: [], flatProps: {} };
  }

  forIn(props, (v, k) => {
    const propChar = CharGenerator.next();
    let found = false;

    if (v.$or) {
      found = true;
      const whereOr = [];
      for (const predicate of v.$or) {
        const tmp = _prepareWhere({ [k]: predicate }, variable);
        whereOr.push(...tmp.where);
        flatProps = { ...flatProps, ...tmp.flatProps };
      }
      where.push(`(${whereOr.join(" OR ")})`);
    } else if (v.$and) {
      found = true;
      const whereAnd = [];
      for (const predicate of v.$and) {
        const tmp = _prepareWhere({ [k]: predicate }, variable);
        whereAnd.push(...tmp.where);
        flatProps = { ...flatProps, ...tmp.flatProps };
      }
      where.push(`(${whereAnd.join(" AND ")})`);
    } else if (v.$not) {
      found = true;
      const tmp = _prepareWhere({ [k]: v.$not }, variable);
      if (tmp.where.length) {
        where.push(`NOT ${tmp.where[0]}`);
        flatProps = { ...flatProps, ...tmp.flatProps };
      }
    } else if (v.$between) {
      found = true;
      if (v.$between.length === 2) {
        const [num1, num2] = v.$between;
        const a = propChar;
        const b = CharGenerator.next();
        where.push(`{${a}} <= ${variable}.${k} <= {${b}}`);
        flatProps[a] = num1 > num2 ? num2 : num1;
        flatProps[b] = num1 > num2 ? num1 : num2;
      }
      // TODO error? flow should actually be able to handle this...
    } else {
      forIn(singlePredicates, (predicateString, predicateKey) => {
        if (v[predicateKey]) {
          found = true;
          where.push(`${variable}.${k} ${predicateString} {${propChar}}`);
          flatProps[propChar] = v[predicateKey];
        }
      });
    }

    if (!found) {
      where.push(`${variable}.${k} = {${propChar}}`);
      flatProps[propChar] = v;
    }
  });

  return { where, flatProps };
}

export function prepareWhere(
  props: any,
  variable: string
): { where: string, flatProps: any } {
  CharGenerator.start("a");
  const { where, flatProps } = _prepareWhere(props, variable);

  if (where.length === 0) {
    return { where: "", flatProps: undefined };
  }

  return { where: "WHERE " + where.join(" AND "), flatProps };
}

export function prepareSet(
  variable: string,
  props: any
): { str: string, newProps: any } {
  const sets = [];
  const newProps: any = {};

  forIn(props, (v, k) => {
    if (k === "guid") return null;
    sets.push(`${variable}.${k}={_u${k}}`);
    newProps[`_u${k}`] = v;
  });

  if (!sets.length) throw new Error(`Nothing to update`);

  return { str: sets.join(", "), newProps };
}
