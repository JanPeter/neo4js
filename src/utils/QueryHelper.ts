import { forIn } from "lodash";
import charGenerator, { CharGenerator } from "./CharGenerator";

const singlePredicates = {
  $sw: "STARTS WITH",
  $ew: "ENDS WITH",
  $contains: "CONTAINS",
  $reg: "=~",

  $eq: "=",
  $in: "IN",

  $gt: ">",
  $gte: ">=",
  $lt: "<",
  $lte: "<=",
};

function _prepareWhere(
  props: any,
  variable: string,
  cg?: CharGenerator
): { where: string[]; flatProps: any } {
  const where = [];
  let flatProps = {};

  if (!props) {
    return { where: [], flatProps: {} };
  }

  forIn(props, (v, k) => {
    const propChar = cg ? cg.next() : charGenerator.next();
    let found = false;

    if (v.$or) {
      found = true;
      const whereOr = [];
      for (const predicate of v.$or) {
        const tmp = _prepareWhere({ [k]: predicate }, variable, cg);
        whereOr.push(...tmp.where);
        flatProps = { ...flatProps, ...tmp.flatProps };
      }
      where.push(`(${whereOr.join(" OR ")})`);
    } else if (v.$and) {
      found = true;
      const whereAnd = [];
      for (const predicate of v.$and) {
        const tmp = _prepareWhere({ [k]: predicate }, variable, cg);
        whereAnd.push(...tmp.where);
        flatProps = { ...flatProps, ...tmp.flatProps };
      }
      where.push(`(${whereAnd.join(" AND ")})`);
    } else if (v.$not) {
      found = true;
      const tmp = _prepareWhere({ [k]: v.$not }, variable, cg);
      if (tmp.where.length) {
        where.push(`NOT ${tmp.where[0]}`);
        flatProps = { ...flatProps, ...tmp.flatProps };
      }
    } else if (v.$between) {
      found = true;
      if (v.$between.length === 2) {
        const [num1, num2] = v.$between;
        const a = propChar;
        const b = cg ? cg.next() : charGenerator.next();
        where.push(`{${a}} <= ${variable}.${k} <= {${b}}`);
        flatProps[a] = num1 > num2 ? num2 : num1;
        flatProps[b] = num1 > num2 ? num1 : num2;
      }
    } else {
      forIn(singlePredicates, (predicateString, predicateKey) => {
        const val = v[predicateKey];
        if (val || typeof val === "number" || typeof val === "string") {
          found = true;
          where.push(`${variable}.${k} ${predicateString} {${propChar}}`);
          flatProps[propChar] = v[predicateKey];
        }
      });
    }

    if (!found) {
      where.push(`${variable}.${k} = {${propChar}}`);
      flatProps[propChar] = v;
    }
  });

  return { where, flatProps };
}

export function prepareWhere(
  properties: any,
  variables: string | string[],
  cg?: CharGenerator
): { where: string | string[]; flatProps: any } {
  if (!cg) charGenerator.start("a");
  let vars: string[] = [];
  let props: any = {};
  if (typeof variables === "string") {
    vars = [variables];
    props[variables] = properties;
  } else if (Array.isArray(variables)) {
    vars = variables;
    props = properties;
  }

  let where = [];
  let flatProps = {};
  vars.forEach(variable => {
    const result = _prepareWhere(props[variable], variable, cg);
    where = where.concat(result.where);
    Object.assign(flatProps, result.flatProps);
  });

  if (where.length === 0) {
    return { where: "", flatProps: undefined };
  }

  return { where: "WHERE " + where.join(" AND "), flatProps };
}

export function prepareSet(
  properties: any,
  variables: string | string[]
): { str: string; newProps: any } {
  const sets = [];
  const newProps: any = {};
  let vars: string[] = [];
  let props: any = {};
  if (typeof variables === "string") {
    vars = [variables];
    props[variables] = properties;
  } else if (Array.isArray(variables)) {
    vars = variables;
    props = properties;
  }

  vars.forEach(variable => {
    forIn(props[variable], (v, k) => {
      if (k === "guid") return null;
      sets.push(`${variable}.${k}={_u_${variable}_${k}}`);
      newProps[`_u_${variable}_${k}`] = v;
    });
  });

  if (!sets.length) throw new Error(`Nothing to update`);

  return { str: sets.join(", "), newProps };
}
