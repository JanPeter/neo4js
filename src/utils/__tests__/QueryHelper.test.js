// @flow

import trineo from "../../index";
import { prepareWhere } from "../QueryHelper";

trineo.init({
  boltUri: "localhost",
  boltPort: 10001,
});

describe("QueryHelper", () => {
  describe("prepareWhere", () => {
    it("should return where equals", () => {
      const where = prepareWhere({ name: "Olaf" }, "a");
      expect(where).toMatchSnapshot();
    });

    it("should return where starts with", () => {
      const where = prepareWhere({ name: { $sw: "O" } }, "a");
      expect(where).toMatchSnapshot();
    });

    it("should return where ends with", () => {
      const where = prepareWhere({ name: { $ew: "f" } }, "a");
      expect(where).toMatchSnapshot();
    });

    it("should return where contains", () => {
      const where = prepareWhere({ name: { $contains: "la" } }, "a");
      expect(where).toMatchSnapshot();
    });

    it("should return where contains", () => {
      const where = prepareWhere({ name: { $contains: "la" } }, "a");
      expect(where).toMatchSnapshot();
    });

    it("should return where regex", () => {
      const where = prepareWhere({ name: { $reg: "Ol.?f" } }, "a");
      expect(where).toMatchSnapshot();
    });

    it("should return where greater than", () => {
      const where = prepareWhere({ age: { $gt: 5 } }, "a");
      expect(where).toMatchSnapshot();
    });

    it("should return where greater than equal", () => {
      const where = prepareWhere({ age: { $gte: 5 } }, "a");
      expect(where).toMatchSnapshot();
    });

    it("should return where less than", () => {
      const where = prepareWhere({ age: { $lt: 5 } }, "a");
      expect(where).toMatchSnapshot();
    });

    it("should return where less than equal", () => {
      const where = prepareWhere({ age: { $lte: 5 } }, "a");
      expect(where).toMatchSnapshot();
    });

    it("should return where starts with or ends with", () => {
      const where = prepareWhere(
        { name: { $or: [{ $sw: "O" }, { $ew: "f" }] } },
        "a"
      );
      expect(where).toMatchSnapshot();
    });

    it("should return where starts with and ends with", () => {
      const where = prepareWhere(
        { name: { $and: [{ $sw: "O" }, { $ew: "f" }] } },
        "a"
      );
      expect(where).toMatchSnapshot();
    });

    it("should return where name in string array", () => {
      const where = prepareWhere({ name: { $in: ["Olaf", "Hubert"] } }, "a");
      expect(where).toMatchSnapshot();
    });

    it("should return where not name equals", () => {
      const where = prepareWhere({ name: { $not: { $eq: "Olaf" } } }, "a");
      expect(where).toMatchSnapshot();
    });

    it("should return where with between", () => {
      const where1 = prepareWhere({ age: { $between: [2, 1] } }, "a");
      const where2 = prepareWhere({ age: { $between: [1, 2] } }, "a");
      expect(where1).toEqual(where2);
      expect(where1).toMatchSnapshot();
    });

    it("should return big where query", () => {
      const where = prepareWhere(
        {
          name: {
            $or: [
              { $eq: "Hans" },
              {
                $and: [{ $sw: "O" }, { $reg: "Ol.?f" }],
              },
            ],
          },
          age: {
            $and: [
              {
                $or: [{ $lt: 5 }, { $gt: 1 }, { $lt: 15 }, { $gt: 10 }],
              },
              { $in: [1, 2, 3, 4, 5, 11, 13] },
            ],
          },
        },
        "a"
      );
      expect(where).toMatchSnapshot();
    });

    it("should return big where query with not", () => {
      const where = prepareWhere(
        {
          name: {
            $not: {
              $or: [
                { $eq: "Hans" },
                {
                  $and: [{ $sw: "O" }, { $reg: "Ol.?f" }],
                },
              ],
            },
          },
          age: {
            $and: [
              {
                $or: [{ $lt: 5 }, { $gt: 1 }, { $lt: 15 }, { $gt: 10 }],
              },
              { $in: [1, 2, 3, 4, 5, 11, 13] },
            ],
          },
        },
        "a"
      );
      expect(where).toMatchSnapshot();
    });
  });
});
