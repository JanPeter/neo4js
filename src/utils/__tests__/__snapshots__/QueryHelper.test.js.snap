// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryHelper prepareWhere should return big where query 1`] = `
Object {
  "flatProps": Object {
    "b": "Hans",
    "d": "O",
    "e": "Ol.?f",
    "h": 5,
    "i": 1,
    "j": 15,
    "k": 10,
    "l": Array [
      1,
      2,
      3,
      4,
      5,
      11,
      13,
    ],
  },
  "where": "WHERE (a.name = {b} OR (a.name STARTS WITH {d} AND a.name =~ {e})) AND ((a.age < {h} OR a.age > {i} OR a.age < {j} OR a.age > {k}) AND a.age IN {l})",
}
`;

exports[`QueryHelper prepareWhere should return big where query with not 1`] = `
Object {
  "flatProps": Object {
    "c": "Hans",
    "e": "O",
    "f": "Ol.?f",
    "i": 5,
    "j": 1,
    "k": 15,
    "l": 10,
    "m": Array [
      1,
      2,
      3,
      4,
      5,
      11,
      13,
    ],
  },
  "where": "WHERE NOT (a.name = {c} OR (a.name STARTS WITH {e} AND a.name =~ {f})) AND ((a.age < {i} OR a.age > {j} OR a.age < {k} OR a.age > {l}) AND a.age IN {m})",
}
`;

exports[`QueryHelper prepareWhere should return where contains 1`] = `
Object {
  "flatProps": Object {
    "a": "la",
  },
  "where": "WHERE a.name CONTAINS {a}",
}
`;

exports[`QueryHelper prepareWhere should return where contains 2`] = `
Object {
  "flatProps": Object {
    "a": "la",
  },
  "where": "WHERE a.name CONTAINS {a}",
}
`;

exports[`QueryHelper prepareWhere should return where ends with 1`] = `
Object {
  "flatProps": Object {
    "a": "f",
  },
  "where": "WHERE a.name ENDS WITH {a}",
}
`;

exports[`QueryHelper prepareWhere should return where equals 1`] = `
Object {
  "flatProps": Object {
    "a": "Olaf",
  },
  "where": "WHERE a.name = {a}",
}
`;

exports[`QueryHelper prepareWhere should return where greater than 1`] = `
Object {
  "flatProps": Object {
    "a": 5,
  },
  "where": "WHERE a.age > {a}",
}
`;

exports[`QueryHelper prepareWhere should return where greater than equal 1`] = `
Object {
  "flatProps": Object {
    "a": 5,
  },
  "where": "WHERE a.age >= {a}",
}
`;

exports[`QueryHelper prepareWhere should return where less than 1`] = `
Object {
  "flatProps": Object {
    "a": 5,
  },
  "where": "WHERE a.age < {a}",
}
`;

exports[`QueryHelper prepareWhere should return where less than equal 1`] = `
Object {
  "flatProps": Object {
    "a": 5,
  },
  "where": "WHERE a.age <= {a}",
}
`;

exports[`QueryHelper prepareWhere should return where name in string array 1`] = `
Object {
  "flatProps": Object {
    "a": Array [
      "Olaf",
      "Hubert",
    ],
  },
  "where": "WHERE a.name IN {a}",
}
`;

exports[`QueryHelper prepareWhere should return where not name equals 1`] = `
Object {
  "flatProps": Object {
    "b": "Olaf",
  },
  "where": "WHERE NOT a.name = {b}",
}
`;

exports[`QueryHelper prepareWhere should return where regex 1`] = `
Object {
  "flatProps": Object {
    "a": "Ol.?f",
  },
  "where": "WHERE a.name =~ {a}",
}
`;

exports[`QueryHelper prepareWhere should return where starts with 1`] = `
Object {
  "flatProps": Object {
    "a": "O",
  },
  "where": "WHERE a.name STARTS WITH {a}",
}
`;

exports[`QueryHelper prepareWhere should return where starts with and ends with 1`] = `
Object {
  "flatProps": Object {
    "b": "O",
    "c": "f",
  },
  "where": "WHERE (a.name STARTS WITH {b} AND a.name ENDS WITH {c})",
}
`;

exports[`QueryHelper prepareWhere should return where starts with or ends with 1`] = `
Object {
  "flatProps": Object {
    "b": "O",
    "c": "f",
  },
  "where": "WHERE (a.name STARTS WITH {b} OR a.name ENDS WITH {c})",
}
`;

exports[`QueryHelper prepareWhere should return where with between 1`] = `
Object {
  "flatProps": Object {
    "a": 1,
    "b": 2,
  },
  "where": "WHERE {a} <= a.age <= {b}",
}
`;
